File: src/typescript/src/builder.ts
/**
 * Fluent builder API for constructing DataFilters filter expressions.
 */

import {
  AndFilter,
  ContainsFilter,
  EndsWithFilter,
  EqualsFilter,
  GreaterThanFilter,
  GreaterThanOrEqualFilter,
  IFilter,
  LessThanFilter,
  LessThanOrEqualFilter,
  NotFilter,
  OrFilter,
  StartsWithFilter,
} from './expressions';

/** Internal helper that provides filter-condition methods for a single field. */
export class FieldBuilder {
  public constructor(
    private readonly parent: FilterBuilder,
    private readonly field: string,
  ) {}

  /** Add an equals condition. */
  public eq(value: unknown): FilterBuilder {
    return this.parent.addFilter(new EqualsFilter(this.field, value));
  }

  /** Add a contains condition. */
  public contains(value: string): FilterBuilder {
    return this.parent.addFilter(new ContainsFilter(this.field, value));
  }

  /** Add a starts-with condition. */
  public startsWith(value: string): FilterBuilder {
    return this.parent.addFilter(new StartsWithFilter(this.field, value));
  }

  /** Add an ends-with condition. */
  public endsWith(value: string): FilterBuilder {
    return this.parent.addFilter(new EndsWithFilter(this.field, value));
  }

  /** Add a greater-than condition. */
  public gt(value: unknown): FilterBuilder {
    return this.parent.addFilter(new GreaterThanFilter(this.field, value));
  }

  /** Add a greater-than-or-equal condition. */
  public gte(value: unknown): FilterBuilder {
    return this.parent.addFilter(new GreaterThanOrEqualFilter(this.field, value));
  }

  /** Add a less-than condition. */
  public lt(value: unknown): FilterBuilder {
    return this.parent.addFilter(new LessThanFilter(this.field, value));
  }

  /** Add a less-than-or-equal condition. */
  public lte(value: unknown): FilterBuilder {
    return this.parent.addFilter(new LessThanOrEqualFilter(this.field, value));
  }

  /** Add a not-equals condition. */
  public not(value: unknown): FilterBuilder {
    return this.parent.addFilter(new NotFilter(new EqualsFilter(this.field, value)));
  }

  /** Add an OR condition matching any of the given values. */
  public or(...values: unknown[]): FilterBuilder {
    const orFilters: IFilter[] = values.map((v) => new EqualsFilter(this.field, v));
    return this.parent.addFilter(new OrFilter(orFilters));
  }
}

/**
 * Fluent builder for constructing compound filter expressions.
 *
 * @example
 * ```ts
 * const filter = new FilterBuilder()
 *   .where('name').contains('bat')
 *   .andWhere('age').gte(18)
 *   .build();
 * ```
 */
export class FilterBuilder {
  private readonly filters: IFilter[] = [];

  /** @internal */
  public addFilter(filter: IFilter): FilterBuilder {
    this.filters.push(filter);
    return this;
  }

  /**
   * Start a filter condition for the given field.
   * @param field - The name of the field to filter on.
   */
  public where(field: string): FieldBuilder {
    return new FieldBuilder(this, field);
  }

  /**
   * Chain an additional AND condition on a new field.
   * @param field - The name of the field for the next condition.
   */
  public andWhere(field: string): FieldBuilder {
    return new FieldBuilder(this, field);
  }

  /**
   * Construct and return the final IFilter.
   *
   * @returns A single IFilter or an AndFilterExpression combining all accumulated conditions.
   * @throws {Error} If no conditions have been added.
   */
  public build(): IFilter {
    if (this.filters.length === 0) {
      throw new Error('No filter conditions have been added.');
    }

    const result: IFilter =
      this.filters.length === 1 ? this.filters[0] : new AndFilter(this.filters);
    return result;
  }
}

File: src/python/datafilters/builder.py
"""Fluent builder API for constructing DataFilters filter expressions."""

from typing import Any, List

from .expressions import (
    AndFilter,
    ContainsFilter,
    EndsWithFilter,
    EqualsFilter,
    GreaterThanFilter,
    GreaterThanOrEqualFilter,
    IFilter,
    LessThanFilter,
    LessThanOrEqualFilter,
    NotFilter,
    OrFilter,
    StartsWithFilter,
)


class FilterBuilder:
    """Fluent builder for constructing compound filter expressions.

    Example usage::

        builder = FilterBuilder()
        f = (builder
             .where("name").contains("bat")
             .and_where("age").gte(18)
             .build())
    """

    def __init__(self) -> None:
        self._filters: List[IFilter] = []

    def where(self, field: str) -> "_FieldBuilder":
        """Start a filter condition for the given field.

        Args:
            field: The name of the field to filter on.

        Returns:
            A ``_FieldBuilder`` scoped to *field*.
        """
        return _FieldBuilder(self, field)

    def _add(self, filter: IFilter) -> "FilterBuilder":
        """Append a filter to the current list."""
        self._filters.append(filter)
        return self

    def and_where(self, field: str) -> "_FieldBuilder":
        """Chain an additional AND condition on a new field.

        Args:
            field: The name of the field for the next condition.

        Returns:
            A ``_FieldBuilder`` scoped to *field*.
        """
        return _FieldBuilder(self, field)

    def build(self) -> IFilter:
        """Construct and return the final IFilter.

        Returns:
            A single IFilter or an AndFilter combining all accumulated conditions.

        Raises:
            ValueError: If no conditions have been added.
        """
        if not self._filters:
            raise ValueError("No filter conditions have been added.")

        if len(self._filters) == 1:
            result = self._filters[0]
        else:
            result = AndFilter(self._filters)

        return result


class _FieldBuilder:
    """Internal helper that provides filter-condition methods for a single field."""

    def __init__(self, parent: FilterBuilder, field: str) -> None:
        self._parent = parent
        self._field = field

    def eq(self, value: Any) -> FilterBuilder:
        """Add an equals condition."""
        return self._parent._add(EqualsFilter(self._field, value))

    def contains(self, value: str) -> FilterBuilder:
        """Add a contains condition."""
        return self._parent._add(ContainsFilter(self._field, value))

    def starts_with(self, value: str) -> FilterBuilder:
        """Add a starts-with condition."""
        return self._parent._add(StartsWithFilter(self._field, value))

    def ends_with(self, value: str) -> FilterBuilder:
        """Add an ends-with condition."""
        return self._parent._add(EndsWithFilter(self._field, value))

    def gt(self, value: Any) -> FilterBuilder:
        """Add a greater-than condition."""
        return self._parent._add(GreaterThanFilter(self._field, value))

    def gte(self, value: Any) -> FilterBuilder:
        """Add a greater-than-or-equal condition."""
        return self._parent._add(GreaterThanOrEqualFilter(self._field, value))

    def lt(self, value: Any) -> FilterBuilder:
        """Add a less-than condition."""
        return self._parent._add(LessThanFilter(self._field, value))

    def lte(self, value: Any) -> FilterBuilder:
        """Add a less-than-or-equal condition."""
        return self._parent._add(LessThanOrEqualFilter(self._field, value))

    def not_(self, value: Any) -> FilterBuilder:
        """Add a not-equals condition."""
        return self._parent._add(NotFilter(EqualsFilter(self._field, value)))

    def or_(self, *values: Any) -> FilterBuilder:
        """Add an OR condition matching any of the given values."""
        or_filters: List[IFilter] = [EqualsFilter(self._field, v) for v in values]
        return self._parent._add(OrFilter(or_filters))
